### PR TITLE
feat(tags): Beginning of v2 backend (setup, create, basic get)

### PR DESCRIPTION
**Pull request by bretthoerner**
This kinda became a beast. Sorry.

To try and simplify review:
* First and foremost, this should have no effect on production. It's a big step toward tagstore v2 backend but legacy should be same-ol, same-ol.
* I was able to trim some fat from the tagstore API, so that's a large portion of the minor changes to legacy backend.
* I needed a tagstore v2 migration so I could run tests.
* I made tagstore `setup` more generic so both backends could use most of the same code (by passing their specific models down).
* The bulk is implementing a large portion of v2 backend and adding tests for it.

### CODE REVIEW

**Comment by getsentry-danger** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">PR is classed as Work in Progress</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">You should update CHANGES due to the size of this PR</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">danger</a>
</p>

File: src/sentry/tagstore/deletions.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+:copyright: (c) 2010-2017 by the Sentry Team, see AUTHORS for more details.
+:license: BSD, see LICENSE for more details.
+"""
+from __future__ import absolute_import, print_function
+
+from sentry.deletions.base import ModelDeletionTask, ModelRelation
+
+
+def tagkeydeletiontask_factory(tagvalue_model, grouptagkey_model, grouptagvalue_model):
```
Comment (mattrobenolt) [inline src/sentry/tagstore/deletions.py]: 🙄

Comment (bretthoerner) [reply src/sentry/tagstore/deletions.py]: get out, matt

File: src/sentry/tagstore/migrations/0001_initial.py [hunk 2]
```diff
@@ -0,0 +1,235 @@
+# -*- coding: utf-8 -*-
```
Comment (mattrobenolt) [inline src/sentry/tagstore/migrations/0001_initial.py]: I don't know how or why, but migrations for nodestore at in `south_migrations` folder. Can we make that happen here too?

File: src/sentry/tagstore/migrations/0001_initial.py [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+from south.db import db
+from south.v2 import SchemaMigration
+from django.db import models
+
+
+class Migration(SchemaMigration):
+
+    # Flag to indicate if this migration is too risky
+    # to run online and needs to be coordinated for offline
+    is_dangerous = False
```
Comment (mattrobenolt) [inline src/sentry/tagstore/migrations/0001_initial.py]: I would argue that we should just set this to True for ourselves even though it's not technically dangerous, just applying this to our existing cluster doesn't need to happen. We'll apply manually to shards.

**Review by mattrobenolt — CHANGES_REQUESTED** [review]
Just my minor nits, but other than that, this looks like hell. Let's ship it.

**Comment by bretthoerner** [comment]
@mattrobenolt fixed your issues

**Review by mattrobenolt — APPROVED** [review]
Punch it. 🍕

**Pull request merged**